### ISSUE TITLE
Precedence of --profile CLI argument vs AWS_PROFILE env var is broken

### ISSUE THREAD

**Issue by mvallaly-rally** [aws/aws-cli#5016]
According to the docs (and in accordance with most unix tools) command line arguments generally override environment variables. https://amzn.to/2tglZeS

This seems to be broken or the docs are wrong.

aws --version
aws-cli/1.17.9 Python/3.8.1 Darwin/19.3.0 botocore/1.14.9

You can easily confirm this behavior via:

AWS_PROFILE=foo aws --profile bar --version

```
$ AWS_PROFILE=foo aws --profile bar --version

Traceback (most recent call last):
  File "/usr/local/bin/aws", line 27, in <module>
    sys.exit(main())
  File "/usr/local/bin/aws", line 23, in main
    return awscli.clidriver.main()
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/awscli/clidriver.py", line 69, in main
    rc = driver.main()
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/awscli/clidriver.py", line 203, in main
    command_table = self._get_command_table()
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/awscli/clidriver.py", line 112, in _get_command_table
    self._command_table = self._build_command_table()
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/awscli/clidriver.py", line 129, in _build_command_table
    self.session.emit('building-command-table.main',
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/session.py", line 674, in emit
    return self._events.emit(event_name, **kwargs)
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/hooks.py", line 356, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/hooks.py", line 228, in emit
    return self._emit(event_name, kwargs)
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/hooks.py", line 211, in _emit
    response = handler(**kwargs)
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/awscli/customizations/preview.py", line 69, in mark_as_preview
    service_name=original_command.service_model.service_name,
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/awscli/clidriver.py", line 328, in service_model
    return self._get_service_model()
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/awscli/clidriver.py", line 345, in _get_service_model
    api_version = self.session.get_config_variable('api_versions').get(
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/session.py", line 237, in get_config_variable
    return self.get_component('config_store').get_config_variable(
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/configprovider.py", line 290, in get_config_variable
    return provider.provide()
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/configprovider.py", line 387, in provide
    value = provider.provide()
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/configprovider.py", line 448, in provide
    scoped_config = self._session.get_scoped_config()
  File "/usr/local/Cellar/awscli/1.17.9/libexec/lib/python3.8/site-packages/botocore/session.py", line 337, in get_scoped_config
    raise ProfileNotFound(profile=profile_name)
botocore.exceptions.ProfileNotFound: The config profile (foo) could not be found
```

Despite the traceback, the last line contains "The config profile (foo) could not be found" this should be using the "bar" profile as specified by the command line argument (--profile bar) passed.

**Comment by benkehoe** [comment]
This is a pretty frustrating bug, as it violates the [expectations for precedence specified in the docs](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html#config-settings-and-precedence). Note also that you can't even set the env var to empty to get around this.
```bash
$ export AWS_PROFILE=
$ aws sts get-caller-identity --profile default
The config profile () could not be found
```

You actually have to `export -n AWS_PROFILE` or `declare +x AWS_PROFILE` to remove the env var entirely.

**Comment by benkehoe** [comment]
I noticed today that it's explicitly called out in the docs that `--profile` should override `AWS_PROFILE`:
![image](https://user-images.githubusercontent.com/1650723/83153727-4a174400-a0cd-11ea-9c1e-d451e9173899.png)
https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html

**Comment by richardhboyd** [comment]
This is going to sound very pedantic, but it is correctly applying the precedence. 

```
$ AWS_PROFILE=burner aws sts get-caller-identity --profile personal
{
    "UserId": "",
    "Account": "[personal profile account id]",
    "Arn": "[personal profile arn]"
}
$ AWS_PROFILE=personal aws sts get-caller-identity --profile burner
{
    "UserId": "AROA6C5M5YBD557ZMNPDM:botocore-session-1590687563",
    "Account": "[burner account id]",
    "Arn": "[burner profile arn]"
}
```

What it appears to be doing is trying to load the profiles in the precedence order (so that the last loaded profile is the highest precedence), if a supplied profile is invalid it will fail when trying to load it. I'll have to do more testing. This could be thorny if one profile specifies, say, a region and the higher precedence one does not so it 'falls back' to the region info in a lower profile. I haven't tested this, but my assumption is that is a behavior that we want. 

All of that being said, it sounds like we need better error messages for invalid profiles.

**Comment by richardhboyd** [comment]
Thinking more, I don't think it will/should 'fall back' to lower profiles since the profiles support the `source_profile` property.

**Comment by benkehoe** [comment]
Yes, I can replicate that behavior in CLI v2, though when I originally encountered it (when helping another user), it seemed like that was not the case. I've also discovered this appears localized to the CLI. This works:
```
AWS_PROFILE=invalid python -c "import boto3, json; print(boto3.Session(profile_name='valid').client('sts').get_caller_identity())"
```

**Comment by kdaily** [comment]
Hi all, this should be resolved with #6066 and available in the AWS CLI versions 1.19.47 and 2.1.36. Thank you!
